wa/workloads/exoplayer/__init__.py

```diff
@@ -0,0 +1,185 @@
+                    self.logger.debug('Downloading {}...'.format(filepath))
+                    urllib.urlretrieve(url, filepath)
+                    self._selected_file = filepath
+                else:
+                    self._selected_file = files[0]
+                    if len(files) > 1:
+                        self.logger.warn('Multiple files for 720p found. Using {}.'
+                                         .format(self._selected_file))
+                        self.logger.warn('Use "filename"parameter instead of '
+                                         '"format" to specify a different file.')
+        return self._selected_file
+
+    def init_resources(self, context):
+        ensure_directory_exists(self.video_directory)
+        self.video_files = defaultdict(list)
+        self.enum_video_files()
+        self._selected_file = None
+
+    def enum_video_files(self):
+        for filename in os.listdir(self.video_directory):
+            for format in self.parameters['format'].allowed_values:
+                if format in filename:
+                    self.video_files[format].append(os.path.join(self.video_directory, filename))
+
+    def validate(self):
+        if (self.format and self.filename) and (self.format != self.parameters['format'].default):
+            raise ConfigError('Ether format *or* filename must be specified; but not both.')
+
+    def initialize(self, context):
+        super(ExoPlayer, self).initialize(context)
+        # Grant app all permissions
+        grant_app_permissions(self.target, self.package)
+
+    def setup(self, context):
+        self.device_video_file = self.target.path.join(self.target.working_directory,
+                                                       os.path.basename(self.host_video_file))
+        if self.force_dependency_push or not self.target.file_exists(self.device_video_file):
+            self.logger.debug('Copying {} to device.'.format(self.host_video_file))
+            self.target.push(self.host_video_file, self.device_video_file, timeout=120)
+
+        super(ExoPlayer, self).setup(context)
+
+        # Prepare logcat monitor
+        self.monitor = self.target.get_logcat_monitor(REGEXPS.values())
+        self.monitor.start()
+
+    def run(self, context):
+        play_cmd = 'am start -a {} -d "file://{}"'.format(self.action, self.device_video_file)
+        self.logger.info(play_cmd)
+        self.target.execute(play_cmd)
+
+        self.monitor.wait_for(REGEXPS['start'])
+        self.logger.info('Playing media file')
+
+        line = self.monitor.wait_for(REGEXPS['duration'])[0]
+        media_duration_s = int(round(float(re.search(REGEXPS['duration'], line)
+                                           .group('duration'))))
+
+        self.logger.info('Media duration is {} seconds'.format(media_duration_s))
+
+        if self.play_duration and self.play_duration < media_duration_s:
```

Comment:
I don't understand this. Why is `play_duration` necessary at all if you're actually waiting for the video to finish?

Reply:
We only wait for the video to finish if `play_duration` is unspecified or the video is longer than that.

Reply:
Right, but I don't  understand why. If `play_duration`  is specified, there are three scenarios:

- `play_duration` == playback duration. `play_duration` is irrelevant.
- `play_duration` < playback duration. `play_duration` is ignored;  `run` returns only actual playback is finished. In this case, the configuration is actively misleading.
- `play_duration` > playback duration. `run` sleeps doing nothing for an arbitrary period of time. Collected measurements are now off  since what was being measured is not actually video playback but the app just sitting there (essentially idle). Here again the setting is misleading as it does not specify the duration of play.

Am I missing something?

Reply:
Oops, when I said 

> We only wait for the video to finish if play_duration is unspecified or the video is longer than that.

I meant if the video is _shorter_ than that. So, the cases are

- `play_duration` == playback duration: irrelevant, yes
- `play_duration` < playback duration: Then we record playing `play_duration` seconds of video (the video is playing while we `time.sleep`).
- `play_duration` > playback duration: Then we record playing the whole video. You are correct that the Parameter is misleading in this case - I think I'll just add a note about it to the description

Reply:
OK, that makes more sense. I also misunderstood how the code worked. From your updated description, the matching of duration returns near the beginning of the workload and does not indicate that playback has finished?

In that case it would make sense; however I would raise an error to force user to adjust the config. (even better, can we query the video file on the host as part of `init_resources`, and raise an error *then* if the configured duration exceeds the video?

Also, please rename to just `duration` to be consistent with other workloads that specify the same parameter.

Reply:
OK, will rename and have it print an error if the video is shorter than `duration` seconds. I won't bother looking at the file on the host - I don't think it would be worth the complexity.

Reply:
I would raise a `ConfigError` rather than just log an error-level message. This is something that needs to be rectified by the user; and we don't want to generate results if the config is wrong.

Reply:
OK